enhancements/20220125-concurrency-control-on-host-disk-operations.md

```diff
@@ -0,0 +1,271 @@
+   was called the second time but the former formatting has not yet finished,
+   this kind of re-entrance problem confuses both operating system and our 
+   custom blockdeivce controller and finally goes stuck in a incorrect state.
+3. All the logic of formatting/partitioning/mounting/provisioning are crammed 
+   into the single `OnBlockDeviceChange` function. That makes debugging and
+   unit-testing nearly impossible. The maintenance cost are still stacking high.
+
+For the first two flaws above, I propose two general concurrency control to
+resolve: lock mechanism and job queue. For the last one, we plan to refactor
+the controller logic to a state-machine-like pattern to enhance maintainability.
+
+[issue #1718]: https://github.com/harvester/harvester/issues/1718
+[feature auto-provision]: https://github.com/harvester/node-disk-manager/pull/17
+
+### User Stories
+
+#### `auto-provision` lots of raw block devices with large volume size
+
+Before this enhancement, user might encounter undefined behavior due to race
+condition of disk formatting re-entrances. Node disk manager sometimes auto 
+heals from this situation but often it fails and leaves blockdevice custom
+resoures in incorrect states.
+
+After this enhancement, the blockdevice custom resource first enters "verb-ing"
+phase. further subsequent changes won't take effect until the disk operations
+completes. The lock and job queue mechanism is nearly unawared from the user
+perspective.
+
+#### Easier troubleshooting and retriggering
+
+Before this enhancement, the status of block device custom resource is entangled
+and difficult for users to interpret. Even developers cannot have a clear view
+on their statuses. It's also hard to retrigger disk operations since one cannot
+simply translate the actual state of the disk from the status of the blockdevice.
+
+After this enhancement, a state-machine-like phase transition helps developers
+see a deterministic path for each phase-to-phase change. Each phase is also a
+single source of true that tells user in what phase the blockdevice are. It can
+also help retrigger disk operations by deleting the blockdevice resource and it
+will run transitions phase-by-phase to the desired phase.
+
+### User Experience In Detail
+
+Basically, user shouldn't aware of this enhancement because this is more like a
+internal refactor than a user-facing feature. However, this enhancement plan to
+1) deprecate `spec.fileSystem.mountPoint` and 2) allow to format raw block
+device directly, so users previous depends on these functionalities will find
+they cannot after this enhancement. Thankfully, these changes won't affect any
+real user experience on disk provisions, since node disk manager won't persist
+mounts for extra disks between boots, and it always re-mounts all extra disk
+after reboot.
+
+### API changes
+
+- :pencil2: Add several new phases to `blockdevice.status.provisionPhase`
+    - :new: `Formatting`
```

Comment:
Don't use `ing` state as indication what's happened. Check the disk status to see if it's formatted/mounted as expected instead.